### ISSUE TITLE
Clustering of items.

### ISSUE THREAD

**Issue by Style87** [almende/vis#242]
I'm looking to add clustering to the timeline as was in it previously. Would you be able to provide me with a starting location or brief comments on what needs to be added? This codebase is rather large and I'm quite lost.

**Comment by AlexDM0** [comment]
Hi,

Over the last year a lot of feature requests have been made. We have just introduced our new website which has a list of the requested features. We have placed this request on that list. 

The list can be found here:
http://visjs.org/featureRequests.html

An explaination of the new system can be found here:
http://visjs.org/blog.html#New\ website\ for\ vis\.js\!

I would like to stress that this does not mean we abandon this request. Discussion here will continue if needed on this feature but we will close it to keep our Github issue page more of a bug-todo list.

Future feature requests will still be made here and then added to the website by us.

Regards,

Alex

**Comment by antferr** [comment]
Why not make available the old clustering capabilities until the new it's ready? It was the best feature!

**Comment by AlexDM0** [comment]
Because the code is completely incompatible.

Regards,

Alex

**Comment by Wuber** [comment]
Hi josdejong and AlexDM0, 

We've been using the chap timeline for a while. It’s a great component, love it and good job!

We’ve recently looked into Vis because we need right-click and multiple select functionality. 
However, we are facing a major hurdle. We actually process large volume of events 
(i.e. thousands of events in very short time frame). The old chap timeline can handle this by clustering events. Since Vis doesn’t support clustering, it simply doesn’t render large events properly in short time frame. 

In summary, the performance (responsiveness and scalability) suffered without clustering. It’s become clear to us that clustering is a must have feature in order for us to move to Vis.

This said, it'd be great to have clustering capability as a high priority feature. Can you please provide  an update or estimate on this feature?

Once again, awesome work, thank you!

**Comment by stevefurey** [comment]
+1 for clustering as well, been waiting for this feature for awhile and need it for all the reasons above.

An update on the progress would be appreciated.

**Comment by josdejong** [comment]
@stevefurey there is no progress on this feature, and we will not be able to pick it up in the near future. Maybe someone from the community is interested in picking this up?

**Comment by ghaliano** [comment]
Any progress on this?

I wrote this example in order to hack the clustering future

It's a data grouping based on timeline scale

`var clusteriser = function(timeline){
	this.timeline = timeline;

	this.getDiffDate = function(date1, date2){
		var second=1000;
		var minute = second*60;
		var hour = minute*60;
		var day = hour*24;
		var month = day*31;
		var year = month*12;

		// Convert both dates to milliseconds
		var diff = date2.getTime() - date1.getTime();

		// Convert back to components and return
		return {
			years: Math.round(diff/year),
			months: Math.round(diff/month),
			days: Math.round(diff/day),
			hours: Math.round(diff/hour),
			minutes: Math.round(diff/minute)
		};
	}

	//Get the scale between two date
	//We can use the scale information from timeline api in order to know how to cluster data
	this.getScale = function(start, end) {	
		var diff = this.getDiffDate(start, end);

		if (diff.months > 12*2) {
			return 'year';
		}
		else if (diff.days >31) {
			return 'month';
		}
		else if (diff.hours >24) {
			return 'day';
		}
		else if (diff.minutes >60) {
			return 'hour';
		}
		else {
			return 'second';
		}
	}

	this.getClusters = function(dataItems){
		var timelineWindow = this.timeline.getWindow();

		var scale = this.getScale(timelineWindow.start, timelineWindow.end);

	    var group = _.toArray(_.groupBy(dataItems, 'group'));;
		var result = {};
		
		var formatByScale = {
			year: 'Y',
			month: 'Y-M',
			weekday: 'Y-M-week-W',
			day: 'Y-M-day-D',
			hour: 'Y-M-D-H',
			second: 'Y-M-D-H-m'
		};

		group.forEach(function(items, i){
			items.forEach(function(item, j){
				var index = moment(item.start).format(formatByScale[scale]) + '-group' + item.group;
				
				if (!result[index]){
					result[index] = {
						count: 1,
						items: [],
						group: item.group,
						start: item.start
					};
				}

				result[index].items.push(item);
				result[index].count++;
			})
		});

		return _.toArray(result);
	}
}`

Usage

`$.getJSON('data.php?limit=1000', function(jsonResponse){
          var groups = jsonResponse['groups'];
          var items = jsonResponse['items'];

          if (!timeline){
            timeline = new vis.Timeline(
              document.getElementById('visualization'), 
              [], 
              groups, 
              {
                template: function (item) {
                  return "<span>"+item.count+" évenement"+((item.count>1)?'s':'')+"</span>"; 

                }
              }
            );

            var clusters = new clusteriser(timeline);
            timeline.setItems(
              clusters.getClusters(items)
            );

            timeline.on("rangechanged", function(){
              timeline.setItems(
                clusters.getClusters(items)
              );
            });
          }
        }
      );`
To see the full example [http://www.developpez.net/forums/blogs/116472-phpiste/b1627/vis-timeline-clustering/](http://www.developpez.net/forums/blogs/116472-phpiste/b1627/vis-timeline-clustering/)

**Comment by sunsetjunks** [comment]
+1 for clustering. The old timeline had a problem with groups, so we transferred to vis-timeline, but we have quite a lot of data to display and the component becomes unusable :-/

**Comment by iislam** [comment]
I was able to cluster data on zoom. 
![time-f-2](https://cloud.githubusercontent.com/assets/4976631/22564825/89475c9c-e94b-11e6-9d30-2ff884306791.gif)

I have written a [blog post here](https://medium.com/@iftekharul/creating-a-timeline-of-events-based-on-customer-feedback-79e6d3d50575#.9ia5pm4kc) about I was able to do it :)

**Comment by yotamberk** [comment]
@iislam very cool! Thank you!

**Comment by mojoaxel** [comment]
@iislam Amazing! Is there any chance that we can use something like this for our [showcases](http://visjs.org/showcase/index.html)?

**Comment by iislam** [comment]
Sure! Let me PM you

**Comment by mojoaxel** [comment]
I added a showcase with #2673 to http://www.iftekhar.me/ibm/ibm-project-timeline/

**Comment by timoteo7** [comment]
+1

**Comment by GluedToTheScreen** [comment]
very nice!

**Comment by yotamberk** [comment]
Closing with thanks to @iffy-i and incorporating #2673 to showcase

**Comment by MeTaNoV** [comment]
the showcase has been done on a codebase which seems rather different, any plan for this request to come back to life?

**Comment by Luiz-N** [comment]
+1 for clustering!!! if I were to decide to use older version (2.9.1) that @iftekharuli used in his blog post what would be lost that the newer versions have? I don't see any immediate performance issues but i'm sure there has been a lot done since then..?

**Comment by jjwallace** [comment]
> I was able to cluster data on zoom.
> ![time-f-2](https://cloud.githubusercontent.com/assets/4976631/22564825/89475c9c-e94b-11e6-9d30-2ff884306791.gif)
> 
> I have written a [blog post here](https://medium.com/@iftekharul/creating-a-timeline-of-events-based-on-customer-feedback-79e6d3d50575#.9ia5pm4kc) about I was able to do it :)

Any link to git hub of your solution?

> @IIslam Amazing! Is there any chance that we can use something like this for our [showcases](http://visjs.org/showcase/index.html)?
> 
> @yotamberk @IIslam Also we need to build an example based on this!

It seems the link has been Corrupted..

**Comment by ryanarnell** [comment]
Yes-- new link to my website- http://www.ryanarnell.com/ibm/ibm-project-timeline/ 
Link to my post- https://medium.com/design-ibm/creating-a-timeline-of-events-based-on-customer-feedback-79e6d3d50575

**Comment by yotamberk** [comment]
Hi @ryanarnell , Great article! There is a typo in your article. 
it's `vis.js` and not `viz.js`. 
In addition, I think it's woth mentioning that vis.js is no longer maintained and `timeline-plus.js` is the new maintained library.

Thanks again and great work!
